hera_cal/abscal.py

```diff
@@ -1291,6 +1295,280 @@ def global_phase_slope_logcal(model, data, antpos, reds=None, solver='linfit', w
+                    break
+            
+            # Apply ratio found
+            for ap in transformed_antpos:
+                transformed_antpos[ap][dim] *= (numerator / denominator)
+
+def complex_phase_abscal(data, model, reds, data_bls, model_bls):
+    """
+    Calculates gains that would absolute calibrate already redundantly-calibrated data.  
+    Only operates one polarization at a time. 
+    
+    Parameters:
+    ----------
+    data : DataContainer or RedDataContainer
+        Dictionary-like container mapping baselines to data visibilities to abscal
+    model : DataContainer or RedDataContainer
+        Dictionary-like container mapping baselines to model visibilities
+    reds : list of lists
+        List of list of redundant baselines tuples like (0, 1, 'ee')
+    data_bls : list of tuples
+        List of baselines in data to use.
+    model_bls : list of tuples
+        List of baselines in model to use. Must correspond the same physical separations as data_bls.
+        
+    Returns:
+    -------
+    meta : dictionary
+        Contains keys for 'Lambda_sol' (degneracy solutions) and 'Z_sol' (solution qualities)
+    delta_gains : dictionary
+        Dictionary mapping antenna keys like (0, 'Jee') to gains of the same shape of the data
+    """
+    # Check that baselines selected are for the same polarization
+    pols = list(set([bl[2] for bls in (data_bls, model_bls) for bl in bls]))
+    assert len(pols) == 1, 'complex_phase_abscal() can only solve for one polarization at a time.'
+    
+    # Get transformed antenna positions and baselines
+    transformed_antpos = redcal.reds_to_antpos(reds)
+    _put_transformed_array_on_integer_grid(transformed_antpos)
+    transformed_b_vecs = np.rint([transformed_antpos[jj] - transformed_antpos[ii] for (ii,jj,pol) in data_bls]).astype(int)
+    
+    # Get number of baselines and times/freqs
+    Ngroups = len(data_bls)
+    Ntimes, Nfreqs = data[data_bls[0]].shape
+
+    # Get normalized data/model ratios and weights
+    normalized_data_model_ratio = np.zeros((Ntimes, Nfreqs, Ngroups), dtype=complex)
+    weights = np.ones((Ntimes, Nfreqs, Ngroups), dtype=float)
+    for nn in range(Ngroups):
+        Vhat_n = data[data_bls[nn]]
+        Vbar_n = model[model_bls[nn]]
+        r_n = Vhat_n / Vbar_n
+        normalized_data_model_ratio[:, :, nn] = r_n / np.abs(r_n)
+        weights[:, :, nn] = np.abs(Vbar_n * Vhat_n)
+
+    # Get solution for degenerate phase slopes
+    weights /= np.mean(weights, axis=-1, keepdims=True)
+    Lambda_sol, Z_sol = _phase_gradient_solution(normalized_data_model_ratio, transformed_b_vecs, weights)
+
+    # turn solution into per-antenna gains
+    phase_angle = {a : np.sum(Lambda_sol * r, axis=-1) for a,r in transformed_antpos.items()}
+    delta_gains = {(a, utils.split_pol(pols[0])[0]): np.exp(1j * (angle)) for a, angle in phase_angle.items()}
+    meta = {'Lambda_sol':  Lambda_sol, 'Z_sol': Z_sol}
```

Comment:
Question for @zacharymartinot: is there anything else that's not too memory intensive that we want to pass through via this meta dict?

Reply:
I added collection of the number of newton's method iterations. We should consider flagging on hitting maxiter.